walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
+      return Arrays.asList();
+    }
+
+    // Get the list of time ranges where the attendees have scheduled events.
+    List<TimeRange> conflicts = findTimeConflicts(events, attendees);
+    return getPossibleMeetingTimes(conflicts, request);
+  }
+
+  /**
+   * Extracts a list of time ranges where any of the given people have events
+   * scheduled. Returns the list sorted by start time in ascending order.
+   */
+  private ImmutableList<TimeRange> findTimeConflicts(
+      Collection<Event> events, Collection<String> people) {
+    return events.stream()
+        .filter(event -> hasCommonAttendees(event, people))
+        .map(Event::getWhen)
+        .sorted(TimeRange.ORDER_BY_START)
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  /**
+   * Returns true if any of the people are attending the event.
+   */
+  private boolean hasCommonAttendees(Event event, Collection<String> people) {
+    Collection<String> attendees = event.getAttendees();
+    return !Collections.disjoint(attendees, people);
+  }
+
+  /**
+   * Gets the list of time ranges when all attendees are free to have
+   * the requested meeting given a list of conflicting time ranges
+   * that is sorted by start time.
+   */
+  private ImmutableList<TimeRange> getPossibleMeetingTimes(
+      List<TimeRange> conflicts, MeetingRequest request) {
+    long meetingDuration = request.getDuration();
+    List<TimeRange> possibleTimes = new ArrayList<>();
+    int freeStart = TimeRange.START_OF_DAY;
+
+    // Add time ranges that are in the gaps between conflicting events.
+    for (int conflictIndex = 0; conflictIndex < conflicts.size(); conflictIndex++) {
+      TimeRange conflict = conflicts.get(conflictIndex);
+      int conflictStart = conflict.start();
+      int conflictEnd = conflict.end();
+
+      // Get the latest end time out of all the time conflicts that overlap with the current one.
+      while (conflictIndex < conflicts.size() - 1
+          && conflict.overlaps(conflicts.get(conflictIndex + 1))) {
+        conflictIndex++;
+        conflict = conflicts.get(conflictIndex);
+
+        // The next overlapping time range may or may not have a later end time.
+        conflictEnd = Math.max(conflictEnd, conflict.end());
+      }
+
+      // Add the free time in between conflicts if it is long enough for the meeting.
+      TimeRange freeTime = TimeRange.fromStartEnd(freeStart, conflictStart, false);
+      if (freeTime.duration() >= meetingDuration) {
+        possibleTimes.add(freeTime);
+      }
+
+      // The start of the next free time range is the end of the latest conflict.
+      freeStart = conflictEnd;
+    }
+
+    // Add the free time between the end of the last conflict and end of day.
+    TimeRange freeTime = TimeRange.fromStartEnd(freeStart, TimeRange.END_OF_DAY, true);
+    if (freeTime.duration() >= meetingDuration) {
+      possibleTimes.add(freeTime);
+    }
+
+    return ImmutableList.copyOf(possibleTimes);
```

Comment:
Instead of creating a copy you can initialize possibleTimes as an ImmutableList using a builder: ImmutableList.Builder<TimeRange> (see https://guava.dev/releases/21.0/api/docs/com/google/common/collect/ImmutableList.Builder.html)